src/vmm/src/resources.rs

```diff
@@ -153,6 +159,67 @@ impl VmResources {
         Ok(resources)
     }
 
+    /// Configures VmResources from the state of the Vmm.
+    pub fn from_vmm(vmm: &Vmm) -> std::result::Result<Self, Error> {
+        let mut resources = Self::default();
+        resources.boot_config = None;
+
+        if let Ok(balloon_config) = vmm.balloon_config() {
+            resources
+                .set_balloon_device(BalloonDeviceConfig::from(balloon_config))
+                .map_err(Error::BalloonDevice)?;
+        }
+
+        vmm.mmio_device_manager.for_each_virtio_device(|virtio_type, _device_id, _device_info, virtio_device| {
+            if virtio_type != TYPE_BLOCK {
+                return Ok(());
+            }
+
+            resources.set_block_device(BlockDeviceConfig::from(
+                virtio_device
+                    .lock()
+                    .expect("Poisoned lock")
+                    .as_any()
+                    .downcast_ref::<Block>()
+                    .unwrap()
+            ))
+            .map_err(Error::BlockDevice)
+        })?;
```

Comment:
The methods here are actually reconstructing the devices. We only need to push the existing ones to the `Builder` objects. You will need to add some helper methods on the Builder objects for this

Reply:
So this piece of feedback is the one that I'm having the hardest time implementing. For the `BlockBuilder` I was thinking about adding a new method like this:

```
pub fn insert_block(&mut self, block: &Block) {
    if block.is_root_device() {
        self.list.push_front(Arc::new(Mutex::new(block.clone())));
    } else {
        self.list.push_back(Arc::new(Mutex::new(block.clone())));
    }
}
```
which would allow using something like the following when setting `VmResources` from a snapshot

```
self.set_block_device(
    virtio_device
        .lock()
        .expect("Poisoned lock")
        .as_any()
        .downcast_ref::<Block>()
        .unwrap()
        .to_owned()
);
```

where `self` is the `VmResources` and the `virtio_device` is one of the arguments of the closure in the `for_each_virtio_device`. The `set_block_device` above would be a method that just calls the `insert_block` on the `BlockBuilder` (the existing `set_block_device` on `VmResources` would be renamed to `set_block_device_from_config` since it uses a `BlockDeviceConfig`). However, this doesn't work because it'd either require that Block satisfies the Clone trait would require a ton of other types inside of Block satisfying the Clone trait which I was thinking wouldn't be the correct approach.

Reply:
What you need to do get a hold of the `Arc<>` part of `Arc<Mutex<Device>>` so that you can clone the reference without also cloning the device.

The problem is that the `MMIODeviceManager::restore` builds devices internally and you never see the `Arc`. Furthermore the `MMIODeviceManager` interface doesn't offer any API to expose the `Arc`: there is only https://github.com/firecracker-microvm/firecracker/blob/83ec483d4254936e1e2fd4adfe94761611a74a93/src/devices/src/bus.rs#L98 which provides only plain reference

So I see two options:
1. Change `MMIODeviceManager::get_device()` and `Bus::get_device()` APIs to return `Option<(u64, Arc<Mutex<dyn BusDevice>>)>`, the you can change the passed function signature in `MMIODeviceManager::for_each_device()` to also take `Arc<Mutex<dyn BusDevice>>` so that within the closure you can clone the `Arc` reference to the device and thus share ownership of it.
2. When restoring devices from states using the `MMIODeviceManager::restore` here https://github.com/firecracker-microvm/firecracker/blob/83ec483d4254936e1e2fd4adfe94761611a74a93/src/vmm/src/builder.rs#L528 you can also provide a function/closure parameter to `MMIODeviceManager::restore` to do auxiliary work on the restored devices (like in your case, also adding them to VmResources). The new signature could look like:
    ```rust
    fn restore(
        constructor_args: Self::ConstructorArgs,
        state: &Self::State,
        for_each_device: Fn(Arc<Mutex<dyn BusDevice>>),
    ) -> Result<Self, Self::Error> {
    ```

I think first option has less cognitive complexity.

Reply:
Sorry again for being slow to respond I got busy over the holidays.

TL;DR: How do I cast a `Arc<Mutex<dyn VirtioDevice>>` into a `Arc<Mutex<Block>>`?

I'm starting to wonder if I just don't have a deep enough understanding of Rust to work on a project as sophisticated as Firecracker. If helping guide me through this issue becomes too burdensome, just let me know if it's best if someone else takes over this bug and I'll completely understand (I can try to contribute again in the future to a simpler problem or when I have a better understanding of Rust/Firecracker).

If you guys are able to continue helping, I'm at a point where I have modified the `MMIODeviceManager::get_device()` and `Bus::get_device()` APIs to return the actual `Arc<Mutex<dyn BusDevice>`, I've modified the closure of `MMIODeviceManager::for_each_virtio_device` to take a `Arc<Mutex<dyn VirtioDevice>>` and I've modified the closure of  `MMIODeviceManager::for_each_device` to take a `Arc<Mutex<dyn BusDevice>>`. The problem I'm having now is that I have a method that looks like 

```rust
pub fn set_block_device(&mut self, block_device: Arc<Mutex<Block>>>)
```
on the `struct VmResources` but the when I clone the `Arc<Mutex<dyn VirtioDevice>>` inside of the `MMIODeviceManager::for_each_device` closure, it can't be passed to `set_block_device` as a `Arc<Mutex<Block>>` since the types don't match. I've tried downcasting the `dyn VirtioDevice` into the `Block` inside of the `Arc<Mutex<dyn VirtioDevice>>` but that seems to cast the entire expression to a `&Block` rather than a `Arc<Mutex<Block>>` which is what I need.

Reply:
Hi @summerlinryan,

> I'm starting to wonder if I just don't have a deep enough understanding of Rust to work on a project as sophisticated as Firecracker.

Don't beat yourself up, this issue is not a straightforward issue. But it is a good learning opportunity! :+1: 

You are right, I don't know if it's possible to clone `Arc<Mutex<dyn Trait>>` into `Arc<Mutex<T: Trait>>`.

The other way around is easily allowed through type coercion (`let t1 = Arc::clone(&struct) as Arc<Mutex<Trait>>;`) because correctness can be easily checked at compile time.

I believe it could be done by implementing `Any` trait for `Mutex<dyn BusDevice>`, `Mutex<Block>`, etc. Then you can use [Arc::downcast](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.downcast) to get what you want.

Having said that, you could also do the simpler thing and go with option 2 from [here](https://github.com/firecracker-microvm/firecracker/pull/2806#discussion_r774466266) with a simple `enum` that gets rid of all dynamic casting:
```rust
enum SharedDeviceTypeWrapper {
    SharedBlock(Arc<Mutex<Block>>),
    SharedNetwork(Arc<Mutex<Net>>),
    SharedBalloon(Arc<Mutex<Balloon>>),
    SharedEtc(Arc<Mutex<Etc>>),
}

fn MMIODeviceManager::restore(
    constructor_args: Self::ConstructorArgs,
    state: &Self::State,
    for_each_restored_device: FnOnce(enum SharedDeviceTypeWrapper),
) -> Result<Self, Self::Error>
{
    // ...
    if let Some(balloon_state) = &state.balloon_device {
        let device = Arc::new(Mutex::new(
            Balloon::restore(
                BalloonConstructorArgs { mem: mem.clone() },
                &balloon_state.device_state,
            )
            .map_err(Error::Balloon)?,
        ));

        // CALL NEW FN HERE:
        for_each_restored_device(SharedDeviceTypeWrapper::SharedBalloon(device.clone()));
        // same code as before after this ^

        restore_helper(
            device.clone(),
            device,
            &balloon_state.device_id,
            &balloon_state.transport_state,
            &balloon_state.mmio_slot,
            constructor_args.event_manager,
        )?;
    }
    // ...
}
```

`pub fn build_microvm_from_snapshot(` could take a `vm_resources: &super::resources::VmResources` reference, and in `build_microvm_from_snapshot` have smth like:
```rust
    // ...
    let register_to_resources = |shared_dev: enum SharedDeviceTypeWrapper| {
        match shared_dev {
            SharedDeviceTypeWrapper::SharedBalloon(b) => vm_resources.register_existing_balloon_device(b),
            ...
        };
    };
    // ...
    vmm.mmio_device_manager =
        MMIODeviceManager::restore(mmio_ctor_args, &microvm_state.device_states, register_to_resources)
            .map_err(MicrovmStateError::RestoreDevices)
            .map_err(RestoreMicrovmState)?;
```

`register_to_resources()` closure in example above could even be a member function of `VmResources` instead of a closure defined in `builder.rs`.

Reply:
I started adding the `for_each_restored_device` function to `MMIODeviceManager::restore` but I noticed that requires adding it to the `Persist` trait which is implemented by a lot of different structs across many different crates. To do so would require either adding a ton of different crate dependencies to allow all of those different crates to know about the new `SharedDeviceTypeWrapper` enum. It would also require adding the new closure argument to all places that currently call the `Persist::restore` method which is quite a few and most of them don't really need to specify the closure anyway since it's primarily just used for setting resources when restoring from a snapshot in `build_microvm_from_snapshot`

Unless I'm misunderstanding the above, then I'm thinking that maybe the next approach to try is to implement Any for `Mutex<dyn VirtioDevice>` like you mentioned, so I'll try that unless you have a workaround to the above problems and still prefer the `for_each_restored_device` approach.

Reply:
I think you can add it to `MMIODevManagerConstructorArgs` [here](https://github.com/firecracker-microvm/firecracker/blob/73ccdfd643fd60b7c22f06fd05765470de3102b2/src/vmm/src/builder.rs#L522) so that you don't have to change the trait (definitely don't want to change that generic trait with this specific param).